File: acc_udp.py
# ──────────────────────────────────────────────
#  ACC UDP Broadcast client
#  Protocol version 4 (ACC 1.6+)
# ──────────────────────────────────────────────
from __future__ import annotations

import socket
import struct
import threading
import time
from enum import IntEnum
from typing import Optional

from data_store import (
    CarEntry, CarRealtime, DataStore, DriverInfo, SessionInfo,
)

PROTOCOL_VERSION = 4


class MsgIn(IntEnum):
    REGISTRATION_RESULT = 1
    REALTIME_UPDATE     = 2
    REALTIME_CAR_UPDATE = 3
    ENTRY_LIST          = 4
    TRACK_DATA          = 5
    ENTRY_LIST_CAR      = 6
    BROADCAST_EVENT     = 7


class MsgOut(IntEnum):
    REGISTER_COMMAND_APPLICATION   = 1
    UNREGISTER_COMMAND_APPLICATION = 9
    REQUEST_ENTRY_LIST             = 10
    REQUEST_TRACK_DATA             = 11


class ACCUDPClient:
    def __init__(
        self,
        data_store: DataStore,
        host: str   = "127.0.0.1",
        port: int   = 9000,
        display_name: str        = "ACC Overlay",
        connection_password: str = "",
        command_password: str    = "",
        update_interval_ms: int  = 250,
    ) -> None:
        self.store                = data_store
        self.host                 = host
        self.port                 = port
        self.display_name         = display_name
        self.connection_password  = connection_password
        self.command_password     = command_password
        self.update_interval_ms   = update_interval_ms

        self.sock: Optional[socket.socket] = None
        self.connection_id = -1
        self._running      = False
        self._thread: Optional[threading.Thread] = None

    # ── Lifecycle ─────────────────────────────

    def start(self) -> None:
        self._running = True
        self._thread  = threading.Thread(target=self._loop, daemon=True, name="acc-udp")
        self._thread.start()

    def stop(self) -> None:
        self._running = False
        if self.sock and self.connection_id >= 0:
            try:
                self.sock.sendto(self._unregister_msg(), (self.host, self.port))
            except OSError:
                pass
        if self.sock:
            self.sock.close()

    # ── Main receive loop ─────────────────────

    def _loop(self) -> None:
        while self._running:
            try:
                self._connect_and_receive()
            except Exception as exc:
                print(f"[ACC UDP] Error: {exc}")
                self.store.set_disconnected()
                time.sleep(5.0)

    def _connect_and_receive(self) -> None:
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.settimeout(5.0)
        sock.bind(("", 0))
        self.sock = sock

        print(f"[ACC UDP] Connecting to {self.host}:{self.port} …")
        sock.sendto(self._register_msg(), (self.host, self.port))

        while self._running:
            try:
                data, _ = sock.recvfrom(65535)
                self._dispatch(data)
            except socket.timeout:
                # Re-send registration so ACC keeps sending us updates
                if self.connection_id < 0:
                    sock.sendto(self._register_msg(), (self.host, self.port))

    # ── Outbound builders ─────────────────────

    @staticmethod
    def _ws(buf: bytearray, s: str) -> None:
        """Write ACC string (uint16 length + UTF-8 bytes)."""
        enc = s.encode("utf-8")
        buf += struct.pack("<H", len(enc))
        buf += enc

    def _register_msg(self) -> bytes:
        buf = bytearray()
        buf.append(MsgOut.REGISTER_COMMAND_APPLICATION)
        buf.append(PROTOCOL_VERSION)
        self._ws(buf, self.display_name)
        self._ws(buf, self.connection_password)
        buf += struct.pack("<i", self.update_interval_ms)
        self._ws(buf, self.command_password)
        return bytes(buf)

    def _unregister_msg(self) -> bytes:
        buf = bytearray()
        buf.append(MsgOut.UNREGISTER_COMMAND_APPLICATION)
        buf += struct.pack("<i", self.connection_id)
        return bytes(buf)

    def _request_entry_list(self) -> None:
        buf = bytearray()
        buf.append(MsgOut.REQUEST_ENTRY_LIST)
        buf += struct.pack("<i", self.connection_id)
        try:
            self.sock.sendto(bytes(buf), (self.host, self.port))
        except OSError:
            pass

    def _request_track_data(self) -> None:
        buf = bytearray()
        buf.append(MsgOut.REQUEST_TRACK_DATA)
        buf += struct.pack("<i", self.connection_id)
        try:
            self.sock.sendto(bytes(buf), (self.host, self.port))
        except OSError:
            pass

    # ── Parser helpers ────────────────────────

    @staticmethod
    def _rs(buf: bytes, off: int) -> tuple[str, int]:
        """Read ACC string → (string, new_offset)."""
        (length,) = struct.unpack_from("<H", buf, off)
        off += 2
        s    = buf[off : off + length].decode("utf-8", errors="replace")
        return s, off + length

    @staticmethod
    def _rl(buf: bytes, off: int) -> tuple[int, int]:
        """Read LapInfo → (lap_time_ms, new_offset)."""
        lap_ms = struct.unpack_from("<i", buf, off)[0]
        off += 4
        off += 2 + 2   # car_index + driver_index
        split_count = buf[off]
        off += 1
        off += split_count * 4   # int32 splits
        off += 4                 # 4 × bool flags
        return lap_ms, off

    # ── Dispatcher ────────────────────────────

    def _dispatch(self, data: bytes) -> None:
        if not data:
            return
        msg_type = data[0]
        off      = 1
        try:
            if   msg_type == MsgIn.REGISTRATION_RESULT:  self._on_registration(data, off)
            elif msg_type == MsgIn.REALTIME_UPDATE:       self._on_realtime_update(data, off)
            elif msg_type == MsgIn.REALTIME_CAR_UPDATE:   self._on_car_update(data, off)
            elif msg_type == MsgIn.ENTRY_LIST:            self._on_entry_list(data, off)
            elif msg_type == MsgIn.ENTRY_LIST_CAR:        self._on_entry_list_car(data, off)
            elif msg_type == MsgIn.TRACK_DATA:            self._on_track_data(data, off)
        except (struct.error, IndexError) as exc:
            print(f"[ACC UDP] Parse error (msg={msg_type}): {exc}")

    # ── Message handlers ──────────────────────

    def _on_registration(self, buf: bytes, off: int) -> None:
        conn_id      = struct.unpack_from("<i", buf, off)[0]; off += 4
        is_read_only = buf[off];                              off += 1
        err_msg, off = self._rs(buf, off)

        if conn_id < 0:
            print(f"[ACC UDP] Registration failed: {err_msg}")
            return

        self.connection_id = conn_id
        self.store.set_connected(conn_id)
        print(f"[ACC UDP] Registered (id={conn_id}, read_only={bool(is_read_only)})")
        self._request_entry_list()
        self._request_track_data()

    def _on_realtime_update(self, buf: bytes, off: int) -> None:
        off += 2 + 2   # event_index, session_index
        session_type = buf[off]; off += 1
        phase        = buf[off]; off += 1
        session_time = struct.unpack_from("<f", buf, off)[0]; off += 4
        session_end  = struct.unpack_from("<f", buf, off)[0]; off += 4
        off += 4       # focused_car_index
        _,  off = self._rs(buf, off)   # active_camera_set
        _,  off = self._rs(buf, off)   # active_camera
        _,  off = self._rs(buf, off)   # current_hud_page
        is_replay = buf[off]; off += 1
        if is_replay:
            off += 8   # replay session_time + remaining_time
        time_of_day = struct.unpack_from("<f", buf, off)[0]; off += 4
        off += 5       # ambient_temp, track_temp, clouds, rain_level, wetness
        best_ms, off = self._rl(buf, off)

        self.store.update_session(SessionInfo(
            session_type    = session_type,
            phase           = phase,
            session_time    = session_time,
            session_end_time= session_end,
            time_of_day     = time_of_day,
            best_session_lap_ms = best_ms,
        ))

    def _on_car_update(self, buf: bytes, off: int) -> None:
        car_index    = struct.unpack_from("<H", buf, off)[0]; off += 2
        driver_index = struct.unpack_from("<H", buf, off)[0]; off += 2
        driver_count = buf[off];                               off += 1
        gear         = struct.unpack_from("<b", buf, off)[0]; off += 1
        off += 4 + 4 + 4   # worldPosX, worldPosY, yaw
        car_location = buf[off]; off += 1
        speed_kmh    = struct.unpack_from("<f", buf, off)[0]; off += 4
        position     = struct.unpack_from("<H", buf, off)[0]; off += 2
        cup_position = struct.unpack_from("<H", buf, off)[0]; off += 2
        track_pos    = struct.unpack_from("<H", buf, off)[0]; off += 2
        spline       = struct.unpack_from("<f", buf, off)[0]; off += 4
        laps         = struct.unpack_from("<H", buf, off)[0]; off += 2
        delta        = struct.unpack_from("<i", buf, off)[0]; off += 4
        best_ms,  off = self._rl(buf, off)
        last_ms,  off = self._rl(buf, off)
        _cur_ms,  off = self._rl(buf, off)

        self.store.update_car_realtime(CarRealtime(
            car_index       = car_index,
            driver_index    = driver_index,
            driver_count    = driver_count,
            gear            = gear,
            speed_kmh       = speed_kmh,
            position        = position,
            cup_position    = cup_position,
            track_position  = track_pos,
            spline_position = spline,
            laps            = laps,
            delta           = delta,
            best_session_lap_ms = best_ms,
            last_lap_ms     = last_ms,
            car_location    = car_location,
        ))

    def _on_entry_list(self, buf: bytes, off: int) -> None:
        off += 4   # connection_id
        # Individual car details will arrive as ENTRY_LIST_CAR packets

    def _on_entry_list_car(self, buf: bytes, off: int) -> None:
        car_index  = struct.unpack_from("<H", buf, off)[0]; off += 2
        car_model  = buf[off];                               off += 1
        team, off  = self._rs(buf, off)
        race_num   = struct.unpack_from("<i", buf, off)[0]; off += 4
        cup_cat    = buf[off];                               off += 1
        cur_driver = struct.unpack_from("<b", buf, off)[0]; off += 1
        drv_count  = buf[off];                               off += 1

        drivers = []
        for _ in range(drv_count):
            fn,  off = self._rs(buf, off)
            ln,  off = self._rs(buf, off)
            sn,  off = self._rs(buf, off)
            cat  = buf[off]; off += 1
            nat, off = self._rs(buf, off)
            drivers.append(DriverInfo(
                first_name  = fn,
                last_name   = ln,
                short_name  = sn,
                category    = cat,
                nationality = nat,
            ))

        nationality = struct.unpack_from("<i", buf, off)[0]

        self.store.update_car_entry(CarEntry(
            car_index            = car_index,
            car_model_type       = car_model,
            team_name            = team,
            race_number          = race_num,
            cup_category         = cup_cat,
            current_driver_index = max(0, cur_driver),
            drivers              = drivers,
            nationality          = nationality,
        ))

    def _on_track_data(self, buf: bytes, off: int) -> None:
        off += 4   # connection_id
        name, off   = self._rs(buf, off)
        off += 4   # track_id
        length_m    = struct.unpack_from("<f", buf, off)[0]
        self.store.update_track(name, length_m)
        print(f"[ACC UDP] Track: {name} ({length_m:.0f} m)")

File: src/acc-udp.js
'use strict';

// ─────────────────────────────────────────────────────────────────────────────
//  ACC Broadcast API v4 – UDP client
//  Reference: Kunos Simulazioni ACC Broadcast SDK (C# reference implementation)
//
//  Message type IDs
const MSG_IN = {
  REGISTRATION_RESULT:  1,
  REALTIME_UPDATE:      2,   // whole-session update
  REALTIME_CAR_UPDATE:  3,   // per-car update  (250 ms)
  ENTRY_LIST:           4,   // car-index list notification
  TRACK_DATA:           5,
  ENTRY_LIST_CAR:       6,   // per-car entry (names, model, …)
};
const MSG_OUT = {
  REGISTER:            1,
  UNREGISTER:          9,
  REQUEST_ENTRY_LIST:  10,
  REQUEST_TRACK_DATA:  11,
};
const PROTOCOL_VERSION = 4;
// ─────────────────────────────────────────────────────────────────────────────

const dgram = require('dgram');

class AccUdpClient {
  constructor(store, options = {}) {
    this.store              = store;
    this.host               = options.host               ?? '127.0.0.1';
    this.port               = options.port               ?? 9000;
    this.displayName        = options.displayName        ?? 'ACC Overlay';
    this.connectionPassword = options.connectionPassword ?? '';
    this.commandPassword    = options.commandPassword    ?? '';
    this.updateIntervalMs   = options.updateIntervalMs   ?? 250;
    this.sock               = null;
    this.connectionId       = -1;
    this._retryTimer        = null;
    this._heartbeatTimer    = null;
    this._entryRefreshTimer = null;
    this._lastPacketMs      = 0;
    this._lastSessionKey    = null;   // 'eventIndex_sessionIndex' — changes on new session
    this._todPrev           = -1;     // previous timeOfDay sample (game seconds)
    this._todRealMs         = 0;      // real timestamp of previous sample
    this._sessClockPrev     = null;   // previous session clock sample (game seconds)
    this._sessClockRealMs   = 0;      // real timestamp of previous session clock sample
    this._tmFromTod         = null;   // preferred weather/day-night multiplier
    this._tmFromSession     = null;   // fallback estimate from session clock
    this._tmTodLastMs       = 0;
  }

  start() { this._connect(); }

  forceRefresh() {
    if (!this.sock) return;
    if (this.connectionId >= 0) {
      this._send(this._buildCmd(MSG_OUT.REQUEST_ENTRY_LIST));
      this._send(this._buildCmd(MSG_OUT.REQUEST_TRACK_DATA));
      console.log('[UDP] Force refresh requested');
      return;
    }
    this._send(this._buildRegister());
    console.log('[UDP] Force refresh requested (re-register)');
  }

  // ── Connection ────────────────────────────────────────────────────────────
  _connect() {
    clearInterval(this._heartbeatTimer);
    clearInterval(this._entryRefreshTimer);
    if (this.sock) { try { this.sock.close(); } catch {} }

    this.sock = dgram.createSocket('udp4');

    this.sock.on('message', (msg) => {
      this._lastPacketMs = Date.now();
      try { this._dispatch(msg); }
      catch (e) {
        this.store.parseErrors++;
        this.store.lastParseErr = e.message;
        console.error('[UDP] dispatch error:', e.message);
      }
    });

    this.sock.on('error', (err) => {
      console.error('[UDP] socket error:', err.message);
      this._scheduleReconnect();
    });

    this.sock.bind(0, () => {
      this._lastPacketMs = Date.now();
      console.log('[UDP] Bound – sending registration…');
      this._send(this._buildRegister());

      // Retry registration every 5 s until ACC responds
      this._retryTimer = setInterval(() => {
        if (this.connectionId < 0) {
          console.log('[UDP] Retrying registration…');
          this._send(this._buildRegister());
        }
      }, 5000);

      // Heartbeat: if connected but no packets for 15 s, ACC probably
      // changed servers — re-register so we get a fresh entry list.
      this._heartbeatTimer = setInterval(() => {
        if (this.connectionId >= 0 && Date.now() - this._lastPacketMs > 15000) {
          console.log('[UDP] Heartbeat timeout — re-registering');
          this.connectionId    = -1;
          this._lastSessionKey = null;
          this.store.setDisconnected();
          this.store.resetForNewServer();
          clearInterval(this._retryTimer);
          this._send(this._buildRegister());
          this._retryTimer = setInterval(() => {
            if (this.connectionId < 0) {
              console.log('[UDP] Retrying registration…');
              this._send(this._buildRegister());
            }
          }, 5000);
        }
      }, 5000);

      // Periodic static entry refresh (late joiners / driver swaps)
      this._entryRefreshTimer = setInterval(() => {
        if (this.connectionId >= 0) {
          this._send(this._buildCmd(MSG_OUT.REQUEST_ENTRY_LIST));
        }
      }, 30000);
    });
  }

  _scheduleReconnect() {
    clearInterval(this._heartbeatTimer);
    clearInterval(this._entryRefreshTimer);
    this.store.setDisconnected();
    this.connectionId = -1;
    clearInterval(this._retryTimer);
    setTimeout(() => this._connect(), 5000);
  }

  _send(buf) {
    this.sock?.send(buf, this.port, this.host,
      (err) => { if (err) console.error('[UDP] send error:', err.message); });
  }

  // ── Outbound builders ────────────────────────────────────────────────────

  /** Write an ACC string: uint16-LE byte length + UTF-8 bytes */
  _ws(parts, s) {
    const enc = Buffer.from(s, 'utf8');
    const len = Buffer.alloc(2); len.writeUInt16LE(enc.length);
    parts.push(len, enc);
  }

  _buildRegister() {
    const parts = [Buffer.from([MSG_OUT.REGISTER, PROTOCOL_VERSION])];
    this._ws(parts, this.displayName);
    this._ws(parts, this.connectionPassword);
    const ms = Buffer.alloc(4); ms.writeInt32LE(this.updateIntervalMs);
    parts.push(ms);
    this._ws(parts, this.commandPassword);
    return Buffer.concat(parts);
  }

  _buildCmd(msgType) {
    const b = Buffer.alloc(5);
    b[0] = msgType;
    b.writeInt32LE(this.connectionId, 1);
    return b;
  }

  // ── Inbound primitives ────────────────────────────────────────────────────

  /** Read ACC string → [string, newOffset] */
  _rs(buf, off) {
    const len = buf.readUInt16LE(off); off += 2;
    return [buf.toString('utf8', off, off + len), off + len];
  }

  /**
   * Read LapInfo → [lapInfo, newOffset]
   * Layout (per Kunos C# SDK):
   *   lapTimeMs    : int32
   *   carIndex     : uint16
   *   driverIndex  : uint16
   *   splitCount   : uint8   (NOT uint16)
   *   splits       : int32 × splitCount
   *   isInvalid    : uint8
   *   isValidForBest: uint8
   *   isOutlap     : uint8
   *   isInlap      : uint8
   */
  _rl(buf, off) {
    const lapMs        = buf.readInt32LE(off);  off += 4;
    /* carIndex */                               off += 2;
    /* driverIndex */                            off += 2;
    const splitCount   = buf[off];               off += 1;
    off += splitCount * 4;
    const isInvalid      = buf[off]; off += 1;
    const isValidForBest = buf[off]; off += 1;
    const isOutlap       = buf[off]; off += 1;
    const isInlap        = buf[off]; off += 1;
    const lapType = isOutlap ? 0 : isInlap ? 2 : 1; // 0=OUT, 1=NORMAL, 2=IN
    return [{ lapMs, isInvalid, isValidForBest, lapType }, off];
  }

  // ── Dispatcher ────────────────────────────────────────────────────────────
  _dispatch(buf) {
    const t   = buf[0];
    const off = 1;       // all parsers start after the type byte
    switch (t) {
      case MSG_IN.REGISTRATION_RESULT: this._onReg(buf, off);      break;
      case MSG_IN.REALTIME_UPDATE:     this._onSession(buf, off);  break;
      case MSG_IN.REALTIME_CAR_UPDATE: this._onCar(buf, off);      break;
      case MSG_IN.ENTRY_LIST_CAR:      this._onCarEntry(buf, off); break;
      case MSG_IN.TRACK_DATA:          this._onTrack(buf, off);    break;
      case MSG_IN.ENTRY_LIST:          /* push notification – ENTRY_LIST_CAR follows automatically */ break;
      default: break;
    }
  }

  // ── Message handlers ─────────────────────────────────────────────────────

  _onReg(buf, off) {
    const id = buf.readInt32LE(off); off += 4;
    const ro = buf[off];             off += 1;
    const [err] = this._rs(buf, off);
    if (id < 0) { console.error('[UDP] Registration failed:', err); return; }

    this.connectionId    = id;
    this._lastSessionKey = null;   // will be set on first REALTIME_UPDATE
    this.store.setConnected(id);
    this.store.resetForNewServer(); // new server → clear both entries and realtimes
    clearInterval(this._retryTimer);
    console.log(`[UDP] Connected  id=${id}  readOnly=${!!ro}`);

    // Request full entry list + track info immediately
    this._send(this._buildCmd(MSG_OUT.REQUEST_ENTRY_LIST));
    this._send(this._buildCmd(MSG_OUT.REQUEST_TRACK_DATA));
  }

  /**
   * REALTIME_UPDATE – whole-session data
   *
   * Layout (v4, offsets from byte 1):
   *   eventIndex      : uint16
   *   sessionIndex    : uint16
   *   sessionType     : uint8
   *   phase           : uint8
   *   sessionTime     : float32  (ms)
   *   sessionEndTime  : float32  (ms)
   *   focusedCarIndex : int32
   *   activeCameraSet : string
   *   activeCamera    : string
   *   currentHudPage  : string
   *   isReplayPlaying : uint8
   *     [if replay] replaySessionTime : float32, replayRemainingTime : float32
   *   timeOfDay       : float32
   *   ambientTemp     : uint8
   *   trackTemp       : uint8
   *   clouds          : uint8   (0–10 scale, skip)
   *   rainLevel       : uint8   (0–10 scale, skip)
   *   wetness         : uint8   (0–10 scale, skip)
   *   weatherTypeNow  : uint8   AccWeatherType: 0=DRY 1=CLOUDS 2=RAIN 3=THUNDERSTORM
   *   weatherIn10min  : uint8
   *   weatherIn30min  : uint8
   *   (remaining bytes: grip forecasts — not parsed here)
   *   NOTE: trackGripStatus is read from acpmf_graphics shared memory (acc-shm.js)
   */
  _onSession(buf, off) {
    const eventIndex   = buf.readUInt16LE(off); off += 2;
    const sessionIndex = buf.readUInt16LE(off); off += 2;
    const sessionKey   = `${eventIndex}_${sessionIndex}`;

    // Detect session change (eventIndex or sessionIndex changed after first update)
    if (this._lastSessionKey !== null && sessionKey !== this._lastSessionKey) {
      console.log(`[UDP] Session changed ${this._lastSessionKey} → ${sessionKey}`);
      // Only clear realtimes — carEntries stays valid because the same cars
      // remain in the server lobby across P/Q/R sessions.
      // Realtimes refill automatically within 250 ms from REALTIME_CAR_UPDATE.
      this.store.resetSessionCars();
    }
    this._lastSessionKey = sessionKey;

    const sessionType    = buf[off]; off += 1;
    const phase          = buf[off]; off += 1;
    const sessionTime    = buf.readFloatLE(off) / 1000; off += 4;   // ms → s
    const sessionEndTime = buf.readFloatLE(off) / 1000; off += 4;   // ms → s
    this._updateTimeMultiplierFromSessionClock(sessionEndTime > 0 ? sessionEndTime : sessionTime);

    // Extended fields (v4) — wrapped in try/catch for older ACC versions
    // NOTE: rain forecast comes from SHM (acc-shm.js), NOT from UDP.
    //       UDP uses a coarser 0-3 AccWeatherType enum; SHM has the accurate
    //       0-5 rainIntensity scale.  We only pull ambientTemp/trackTemp here.
    let focusedCarIndex = -1;
    let ambientTemp, trackTemp;
    let timeOfDay;
    try {
      focusedCarIndex = buf.readInt32LE(off); off += 4;
      let _cam1, _cam2, _hud;
      [_cam1, off] = this._rs(buf, off);   // activeCameraSet
      [_cam2, off] = this._rs(buf, off);   // activeCamera
      [_hud,  off] = this._rs(buf, off);   // currentHudPage
      const isReplay = buf[off]; off += 1;
      if (isReplay) off += 8;              // replaySessionTime + replayRemainingTime (2×float32)
      timeOfDay = buf.readFloatLE(off); off += 4;  // seconds since midnight (game time)
      this._updateTimeMultiplier(timeOfDay);
      ambientTemp = buf[off]; off += 1;
      trackTemp   = buf[off];
    } catch { /* older ACC — extended fields unavailable */ }

    this.store.updateSession({
      sessionType, phase, sessionTime, sessionEndTime,
      focusedCarIndex,
      ambientTemp, trackTemp,
      eventIndex, sessionIndex, sessionKey,
    });
  }

  /** Calculate game-time multiplier from timeOfDay progression */
  _updateTimeMultiplier(tod) {
    const now = Date.now();
    if (!(tod > 0)) return;
    if (this._todPrev >= 0) {
      if (Math.abs(tod - this._todPrev) < 0.001) return; // ignore unchanged coarse samples
      const realDeltaS = (now - this._todRealMs) / 1000;
      let gameDelta = tod - this._todPrev;
      // Handle midnight wrap (86400 = seconds in a day)
      if (gameDelta < -43200) gameDelta += 86400;
      if (realDeltaS > 0.2 && realDeltaS < 5 && gameDelta > 0) {
        const mult = gameDelta / realDeltaS;
        if (!Number.isFinite(mult) || mult < 0.1 || mult > 30) {
          this._todPrev = tod;
          this._todRealMs = now;
          return;
        }
        const prev = this._tmFromTod ?? mult;
        this._tmFromTod = prev * 0.8 + mult * 0.2;
        this._tmTodLastMs = now;
        this._commitTimeMultiplier();
      }
    }
    this._todPrev = tod;
    this._todRealMs = now;
  }

  /** Fallback multiplier estimate from session clock progression (works even if timeOfDay is unreliable). */
  _updateTimeMultiplierFromSessionClock(clockS) {
    const now = Date.now();
    if (!(clockS >= 0)) return;
    if (this._tmTodLastMs > 0 && (now - this._tmTodLastMs) < 15000) {
      this._sessClockPrev = clockS;
      this._sessClockRealMs = now;
      return;
    }
    if (this._sessClockPrev != null && this._sessClockRealMs > 0) {
      const realDeltaS = (now - this._sessClockRealMs) / 1000;
      const gameDeltaAbs = Math.abs(clockS - this._sessClockPrev);
      if (realDeltaS > 0.2 && realDeltaS < 5 && gameDeltaAbs > 0.001) {
        const mult = gameDeltaAbs / realDeltaS;
        if (Number.isFinite(mult) && mult >= 0.1 && mult <= 30) {
          const prev = this._tmFromSession ?? mult;
          this._tmFromSession = prev * 0.8 + mult * 0.2;
          this._commitTimeMultiplier();
        }
      }
    }
    this._sessClockPrev = clockS;
    this._sessClockRealMs = now;
  }

  _commitTimeMultiplier() {
    const tm = this._tmFromTod ?? this._tmFromSession;
    if (Number.isFinite(tm) && tm > 0) this.store.session.timeMultiplier = tm;
  }

  /**
   * REALTIME_CAR_UPDATE – per-car data every ~250 ms
   *
   * Layout (v4):
   *   carIndex      : uint16
   *   driverIndex   : uint16
   *   driverCount   : uint8
   *   gear          : uint8   (0=R, 1=N, 2=1st, …)
   *   worldPosX     : float32
   *   worldPosY     : float32
   *   yaw           : float32
   *   carLocation   : uint8   (0=NONE,1=TRACK,2=PITLANE,3=PITENTRY,4=PITEXIT)
   *   speedKmh      : uint16  (integer km/h – NOT float!)
   *   position      : uint16  (overall race position, 1-based)
   *   cupPosition   : uint16
   *   trackPosition : uint16
   *   splinePosition: float32 (0.0–1.0 along track)
   *   laps          : uint16
   *   delta         : int32
   *   (lap times are read from acpmf_graphics SHM — see acc-shm.js)
   */
  _onCar(buf, off) {
    try {
      const carIndex       = buf.readUInt16LE(off); off += 2;
      const driverIndex    = buf.readUInt16LE(off); off += 2;
      const driverCount    = buf[off];              off += 1;
      const gear           = buf[off];              off += 1;
      off += 12;   // worldPosX + worldPosY + yaw  (3 × float32)
      const carLocation    = buf[off];              off += 1;
      const speedKmh       = buf.readUInt16LE(off); off += 2;   // ← uint16, NOT float32
      const position       = buf.readUInt16LE(off); off += 2;
      const cupPosition    = buf.readUInt16LE(off); off += 2;
      const trackPosition  = buf.readUInt16LE(off); off += 2;
      const splinePosition = buf.readFloatLE(off);  off += 4;
      const laps           = buf.readUInt16LE(off); off += 2;
      const delta          = buf.readInt32LE(off);  off += 4;

      // Read LapInfo: bestSessionLap first, then lastLap
      // Each try is independent so a failure on lastLap doesn't lose bestSessionLap
      let bestSessionLapMs = -1;
      let lastLapMs        = -1;
      let lastLapValidForBest = undefined;
      let lastLapTypeRaw      = undefined;
      let off2 = off;
      try {
        const [bestLap, newOff] = this._rl(buf, off);
        off2 = newOff;
        if (bestLap && typeof bestLap.lapMs === 'number') {
          bestSessionLapMs = bestLap.lapMs;
        }
      } catch {}
      try {
        const [lastLap] = this._rl(buf, off2);
        if (lastLap && typeof lastLap.lapMs === 'number') {
          lastLapMs = lastLap.lapMs;
          // Treat "valid for best" and "not invalid" as our validity flag
          if (typeof lastLap.isInvalid === 'number' && typeof lastLap.isValidForBest === 'number') {
            lastLapValidForBest = !lastLap.isInvalid && !!lastLap.isValidForBest;
          }
          if (typeof lastLap.lapType === 'number') {
            lastLapTypeRaw = lastLap.lapType;
          }
        }
      } catch {}


      this.store.updateCarRealtime({
        carIndex, driverIndex, driverCount, gear,
        speedKmh, position, cupPosition, trackPosition,
        splinePosition, laps, delta, carLocation,
        bestSessionLapMs, lastLapMs,
        lastLapValidForBest, lastLapTypeRaw,
      });
    } catch (e) {
      this.store.parseErrors++;
      this.store.lastParseErr = 'car:' + e.message;
    }
  }

  /**
   * ENTRY_LIST_CAR – static car/driver info (sent once after REQUEST_ENTRY_LIST)
   *
   * Layout (v4):
   *   carIndex           : uint16
   *   carModelType       : uint8
   *   teamName           : string
   *   raceNumber         : int32
   *   cupCategory        : uint8
   *   currentDriverIndex : uint8
   *   carNationality     : uint16  ← v4 addition (skip it)
   *   driversCount       : uint8
   *   [ for each driver ]
   *     firstName        : string
   *     lastName         : string
   *     shortName        : string
   *     driverCategory   : uint8
   *     nationality      : uint16  ← uint16, NOT a string!
   */
  _onCarEntry(buf, off) {
    try {
      const carIndex   = buf.readUInt16LE(off); off += 2;
      const carModel   = buf[off];              off += 1;
      let teamName; [teamName, off] = this._rs(buf, off);
      const raceNumber = buf.readInt32LE(off);  off += 4;
      const cupCat     = buf[off];              off += 1;
      const curDriver  = buf[off];              off += 1;
      off += 2;   // carNationality uint16 (v4) – not needed
      const drvCount   = buf[off];              off += 1;

      const drivers = [];
      for (let i = 0; i < drvCount; i++) {
        let fn, ln, sn;
        [fn, off] = this._rs(buf, off);
        [ln, off] = this._rs(buf, off);
        [sn, off] = this._rs(buf, off);
        const cat = buf[off];              off += 1;
        const nat = buf.readUInt16LE(off); off += 2;   // nationality uint16
        drivers.push({ firstName: fn, lastName: ln, shortName: sn, category: cat, nationality: nat });
      }

      this.store.updateCarEntry({
        carIndex, carModelType: carModel, teamName, raceNumber,
        cupCategory: cupCat,
        currentDriverIndex: Math.min(curDriver, Math.max(0, drvCount - 1)),
        drivers,
      });
    } catch (e) {
      this.store.parseErrors++;
      this.store.lastParseErr = 'entry:' + e.message;
      console.error('[UDP] _onCarEntry parse error:', e.message);
    }
  }

  /**
   * TRACK_DATA
   *
   * Layout:
   *   connectionId  : int32
   *   trackName     : string
   *   trackId       : int32
   *   trackLengthM  : float32
   *   … (camera sets – not needed)
   */
  _onTrack(buf, off) {
    off += 4;   // skip connectionId
    let name; [name, off] = this._rs(buf, off);
    off += 4;   // skip trackId
    const lengthM = buf.readFloatLE(off);
    this.store.updateTrack(name, lengthM);
    console.log(`[UDP] Track: ${name}  (${Math.round(lengthM)} m)`);
  }
}

module.exports = AccUdpClient;
